// althea-frontend/App.jsx
import React, { useEffect, useMemo, useState } from 'react';
import Header from './components/Header.jsx';
import Footer from './components/Footer.jsx';
import './styles/App.css';

import Home from './pages/Home/Home.jsx';
import Category from './pages/Category/Category.jsx';
import Product from './pages/Product/Product.jsx';
import Search from './pages/Search/Search.jsx';
import Cart from './pages/Cart/Cart.jsx';
import Checkout from './pages/Checkout/Checkout.jsx';
import Confirmation from './pages/Confirmation/Confirmation.jsx';
import Register from './pages/Register/Register.jsx';
import Login from './pages/Login/Login.jsx';
import Account from './pages/Account/Account.jsx';
import OrderHistory from './pages/OrderHistory/OrderHistory.jsx';
import Contact from './pages/Contact/Contact.jsx';
import Admin from './pages/Admin/Admin.jsx';
import { useLocalStorage } from './hooks/useLocalStorage.js';
import {
  initialCart,
  initialCategories,
  initialHomeContent,
  initialOrders,
  initialProducts,
  initialSession,
  initialUser,
} from './data/mockData.js';
import {
  buildCartDetails,
  buildRelatedProducts,
  computeCartSummary,
  createOrderId,
  searchProducts,
  sortProductsForCategory,
} from './utils/storefront.js';

const navItems = [
  { label: 'Accueil', path: '/' },
  { label: 'Catalogue', path: '/category' },
  { label: 'Recherche', path: '/search' },
  { label: 'Commandes', path: '/orders' },
  { label: 'Contact', path: '/contact' },
  { label: 'Admin', path: '/admin' },
];

const initialSearchState = {
  query: '',
  description: '',
  technical: '',
  categoryId: 'all',
  availableOnly: false,
  minPrice: '',
  maxPrice: '',
  sortBy: 'relevance',
  sortDirection: 'asc',
};

function parseHashLocation() {
  const normalizedHash = window.location.hash.replace(/^#/, '').trim() || '/';
  const [path, queryString = ''] = normalizedHash.split('?');

  return {
    path: path || '/',
    params: new URLSearchParams(queryString),
  };
}

function buildHash(path, params = {}) {
  const searchParams = new URLSearchParams();

  Object.entries(params).forEach(([key, value]) => {
    if (value !== undefined && value !== null && value !== '') {
      searchParams.set(key, value);
    }
  });

  const queryString = searchParams.toString();
  return `#${path}${queryString ? `?${queryString}` : ''}`;
}

function getPasswordValidation(password) {
  return (
    password.length >= 8 &&
    /[A-Z]/.test(password) &&
    /[a-z]/.test(password) &&
    /\d/.test(password) &&
    /[^A-Za-z0-9]/.test(password)
  );
}

export default function App() {
  const [location, setLocation] = useState(parseHashLocation());
  const [categories, setCategories] = useLocalStorage('althea-categories', initialCategories);
  const [products, setProducts] = useLocalStorage('althea-products', initialProducts);
  const [homeContent, setHomeContent] = useLocalStorage('althea-home-content', initialHomeContent);
  const [session, setSession] = useLocalStorage('althea-session', initialSession);
  const [userProfile, setUserProfile] = useLocalStorage('althea-user-profile', initialUser);
  const [cartItems, setCartItems] = useLocalStorage('althea-cart', initialCart);
  const [orders, setOrders] = useLocalStorage('althea-orders', initialOrders);
  const [lastOrderId, setLastOrderId] = useLocalStorage('althea-last-order-id', initialOrders[0]?.id ?? null);
  const [searchState, setSearchState] = useLocalStorage('althea-search-state', initialSearchState);

  useEffect(() => {
    if (!window.location.hash) {
      window.location.hash = '#/';
    }

    const onHashChange = () => setLocation(parseHashLocation());
    window.addEventListener('hashchange', onHashChange);

    return () => window.removeEventListener('hashchange', onHashChange);
  }, []);

  useEffect(() => {
    if (location.path === '/search') {
      const queryInHash = location.params.get('q') || '';
      if (queryInHash && queryInHash !== searchState.query) {
        setSearchState((previous) => ({ ...previous, query: queryInHash }));
      }
    }
  }, [location.params, location.path, searchState.query, setSearchState]);

  const sortedCategories = useMemo(
    () => [...categories].sort((left, right) => left.displayOrder - right.displayOrder),
    [categories],
  );

  const activeCategorySlug = location.params.get('category') || sortedCategories[0]?.slug;
  const activeCategory = sortedCategories.find((category) => category.slug === activeCategorySlug) || sortedCategories[0];

  const categoryProducts = useMemo(() => {
    if (!activeCategory) {
      return [];
    }

    return sortProductsForCategory(products.filter((product) => product.categoryId === activeCategory.id));
  }, [activeCategory, products]);

  const selectedProductSlug = location.params.get('product') || products[0]?.slug;
  const selectedProduct = products.find((product) => product.slug === selectedProductSlug) || products[0] || null;

  const relatedProducts = useMemo(() => {
    if (!selectedProduct) {
      return [];
    }

    return buildRelatedProducts(products, selectedProduct, 6);
  }, [products, selectedProduct]);

  const featuredProducts = useMemo(
    () => [...products].filter((product) => product.featuredRank > 0).sort((left, right) => left.featuredRank - right.featuredRank),
    [products],
  );

  const searchResults = useMemo(() => searchProducts(products, searchState), [products, searchState]);
  const cartDetails = useMemo(() => buildCartDetails(cartItems, products), [cartItems, products]);
  const cartSummary = useMemo(() => computeCartSummary(cartDetails), [cartDetails]);
  const currentOrder = orders.find((order) => order.id === (location.params.get('order') || lastOrderId)) || orders[0] || null;

  const navigate = (path, params = {}) => {
    window.location.hash = buildHash(path, params);
  };

  const handleCategoryNavigation = (categorySlug) => navigate('/category', { category: categorySlug });

  const handleProductNavigation = (productSlug) => navigate('/product', { product: productSlug });

  const handleHeaderSearch = (query) => {
    setSearchState((previous) => ({ ...previous, query }));
    navigate('/search', { q: query });
  };

  const handleAddToCart = (productId, quantity = 1) => {
    setCartItems((previous) => {
      const existingItem = previous.find((item) => item.productId === productId);

      if (existingItem) {
        return previous.map((item) =>
          item.productId === productId ? { ...item, quantity: item.quantity + quantity } : item,
        );
      }

      return [...previous, { productId, quantity }];
    });
  };

  const handleUpdateCartQuantity = (productId, quantity) => {
    if (quantity <= 0) {
      setCartItems((previous) => previous.filter((item) => item.productId !== productId));
      return;
    }

    setCartItems((previous) =>
      previous.map((item) => (item.productId === productId ? { ...item, quantity } : item)),
    );
  };

  const handleRemoveCartItem = (productId) => {
    setCartItems((previous) => previous.filter((item) => item.productId !== productId));
  };

  const handleLogin = ({ email, password }) => {
    if (!email || !password) {
      return { success: false, message: 'Veuillez renseigner votre e-mail et votre mot de passe.' };
    }

    if (password.length < 8) {
      return { success: false, message: 'Mot de passe incorrect. Pensez à utiliser le lien “mot de passe oublié”.' };
    }

    setSession({ isAuthenticated: true, role: 'customer' });
    setUserProfile((previous) => ({ ...previous, email }));

    return { success: true, message: 'Connexion réussie.' };
  };

  const handleRegister = ({ firstName, lastName, email, password, company }) => {
    if (!firstName || !lastName || !email) {
      return { success: false, message: 'Tous les champs obligatoires doivent être complétés.' };
    }

    if (!getPasswordValidation(password)) {
      return {
        success: false,
        message: 'Le mot de passe doit contenir 8 caractères, une majuscule, une minuscule, un chiffre et un caractère spécial.',
      };
    }

    setUserProfile((previous) => ({
      ...previous,
      firstName,
      lastName,
      email,
      company,
      verified: false,
    }));
    setSession({ isAuthenticated: true, role: 'customer' });

    return {
      success: true,
      message: 'Compte créé. La confirmation e-mail sera à brancher côté backend.',
    };
  };

  const handleSaveAccount = (nextProfile) => {
    setUserProfile((previous) => ({
      ...previous,
      ...nextProfile,
    }));
  };

  const handlePlaceOrder = ({ billingAddress, paymentDetails }) => {
    if (cartSummary.unavailableCount > 0) {
      return { success: false, message: 'Retirez les produits indisponibles avant validation.' };
    }

    if (cartDetails.length === 0) {
      return { success: false, message: 'Votre panier est vide.' };
    }

    const nextOrder = {
      id: createOrderId(),
      createdAt: new Date().toISOString().slice(0, 10),
      status: 'À confirmer côté paiement',
      totalCents: cartSummary.totalCents,
      items: cartDetails.map((item) => ({ productId: item.productId, quantity: item.quantity })),
      billingAddress,
      paymentSummary: `${paymentDetails.cardholderName} •••• ${paymentDetails.cardNumber.slice(-4)}`,
    };

    setOrders((previous) => [nextOrder, ...previous]);
    setLastOrderId(nextOrder.id);
    setCartItems([]);
    setUserProfile((previous) => ({
      ...previous,
      addresses: previous.addresses?.length ? previous.addresses : [billingAddress],
      paymentMethods: previous.paymentMethods?.length
        ? previous.paymentMethods
        : [
            {
              id: `pm-${Date.now()}`,
              label: 'Carte enregistrée',
              cardholderName: paymentDetails.cardholderName,
              last4: paymentDetails.cardNumber.slice(-4),
              expiry: paymentDetails.expiry,
            },
          ],
    }));

    navigate('/confirmation', { order: nextOrder.id });
    return { success: true, message: 'Commande validée.' };
  };

  const handleMoveCarouselSlide = (slideId, direction) => {
    setHomeContent((previous) => {
      const slides = [...previous.carousel];
      const index = slides.findIndex((slide) => slide.id === slideId);
      const targetIndex = direction === 'up' ? index - 1 : index + 1;

      if (index < 0 || targetIndex < 0 || targetIndex >= slides.length) {
        return previous;
      }

      [slides[index], slides[targetIndex]] = [slides[targetIndex], slides[index]];
      return { ...previous, carousel: slides };
    });
  };

  const handleToggleProductPriority = (productId) => {
    setProducts((previous) => {
      const maxPriority = Math.max(0, ...previous.map((product) => product.priorityRank || 0));
      return previous.map((product) =>
        product.id === productId
          ? { ...product, priorityRank: product.priorityRank > 0 ? 0 : maxPriority + 1 }
          : product,
      );
    });
  };

  const handleToggleProductAvailability = (productId) => {
    setProducts((previous) =>
      previous.map((product) =>
        product.id === productId
          ? { ...product, availableStock: product.availableStock > 0 ? 0 : 10 }
          : product,
      ),
    );
  };

  const handleToggleFeatured = (productId) => {
    setProducts((previous) => {
      const maxFeatured = Math.max(0, ...previous.map((product) => product.featuredRank || 0));
      return previous.map((product) =>
        product.id === productId
          ? { ...product, featuredRank: product.featuredRank > 0 ? 0 : maxFeatured + 1 }
          : product,
      );
    });
  };

  const handleSetCategoryOrder = (categoryId, displayOrder) => {
    setCategories((previous) =>
      previous.map((category) =>
        category.id === categoryId ? { ...category, displayOrder: Number(displayOrder) } : category,
      ),
    );
  };

  const handleUpdateHomeMessage = (fixedMessage) => {
    setHomeContent((previous) => ({ ...previous, fixedMessage }));
  };

  const cartCount = cartItems.reduce((sum, item) => sum + item.quantity, 0);

  const currentPage = useMemo(() => {
    switch (location.path) {
      case '/category':
        return (
          <Category
            categories={sortedCategories}
            activeCategory={activeCategory}
            products={categoryProducts}
            onSelectCategory={handleCategoryNavigation}
            onOpenProduct={handleProductNavigation}
          />
        );
      case '/product':
        return (
          <Product
            product={selectedProduct}
            relatedProducts={relatedProducts}
            onAddToCart={handleAddToCart}
            onBuyNow={(productId) => {
              handleAddToCart(productId, 1);
              navigate('/checkout');
            }}
            onOpenProduct={handleProductNavigation}
          />
        );
      case '/search':
        return (
          <Search
            categories={sortedCategories}
            criteria={searchState}
            onChangeCriteria={setSearchState}
            results={searchResults}
            onOpenProduct={handleProductNavigation}
          />
        );
      case '/cart':
        return (
          <Cart
            items={cartDetails}
            summary={cartSummary}
            isAuthenticated={session.isAuthenticated}
            onUpdateQuantity={handleUpdateCartQuantity}
            onRemoveItem={handleRemoveCartItem}
            onNavigate={navigate}
          />
        );
      case '/checkout':
        return (
          <Checkout
            cartItems={cartDetails}
            summary={cartSummary}
            user={userProfile}
            session={session}
            onNavigate={navigate}
            onLogin={handleLogin}
            onRegister={handleRegister}
            onPlaceOrder={handlePlaceOrder}
          />
        );
      case '/confirmation':
        return <Confirmation order={currentOrder} products={products} onNavigate={navigate} />;
      case '/register':
        return <Register onRegister={handleRegister} onNavigate={navigate} />;
      case '/login':
        return <Login onLogin={handleLogin} onNavigate={navigate} />;
      case '/account':
        return (
          <Account
            user={userProfile}
            session={session}
            orders={orders}
            onSave={handleSaveAccount}
            onNavigate={navigate}
          />
        );
      case '/orders':
        return <OrderHistory orders={orders} products={products} onNavigate={navigate} />;
      case '/contact':
        return <Contact onNavigate={navigate} />;
      case '/admin':
        return (
          <Admin
            homeContent={homeContent}
            categories={sortedCategories}
            products={products}
            orders={orders}
            onUpdateHomeMessage={handleUpdateHomeMessage}
            onMoveCarouselSlide={handleMoveCarouselSlide}
            onToggleProductPriority={handleToggleProductPriority}
            onToggleProductAvailability={handleToggleProductAvailability}
            onToggleFeatured={handleToggleFeatured}
            onSetCategoryOrder={handleSetCategoryOrder}
            onOpenProduct={handleProductNavigation}
          />
        );
      case '/':
      default:
        return (
          <Home
            homeContent={homeContent}
            categories={sortedCategories}
            featuredProducts={featuredProducts}
            onOpenCategory={handleCategoryNavigation}
            onOpenProduct={handleProductNavigation}
            onNavigate={navigate}
          />
        );
    }
  }, [
    activeCategory,
    cartDetails,
    cartSummary,
    categoryProducts,
    currentOrder,
    featuredProducts,
    homeContent,
    location.path,
    orders,
    products,
    relatedProducts,
    searchResults,
    searchState,
    session,
    sortedCategories,
    userProfile,
  ]);

  return (
    <div className="app-shell">
      <Header
        navItems={navItems}
        currentPath={location.path}
        cartCount={cartCount}
        searchValue={searchState.query}
        isAuthenticated={session.isAuthenticated}
        onNavigate={navigate}
        onSearchSubmit={handleHeaderSearch}
      />

      <main className="app-main">{currentPage}</main>

      <Footer onNavigate={navigate} />
    </div>
  );
}
